File: FateGrandAutomata/Screenshot/RootScreenshotService.cs
using System.IO;
using System.Text;
using Android.OS;
using CoreAutomata;
using Java.Lang;
using Org.Opencv.Core;
using Org.Opencv.Imgproc;
using Process = Java.Lang.Process;

namespace FateGrandAutomata
{
    public class RootScreenshotService : IScreenshotService
    {
        readonly Process _superUser;
        readonly StreamWriter _superUserStreamWriter;
        byte[] _buffer;
        Mat _rootLoadMat, _rootConvertMat = new Mat();

        readonly string _imgPath;

        public RootScreenshotService()
        {
            try
            {
                _superUser = Runtime.GetRuntime().Exec("su", null, null);
                _superUserStreamWriter = new StreamWriter(_superUser.OutputStream, Encoding.ASCII);
            }
            catch (Exception e)
            {
                throw new System.Exception("Failed to get Root permission", e);
            }


            _imgPath = Path.Combine(AutomataApi.StorageDir, "sshot.raw");
        }

        public IPattern TakeScreenshot()
        {
            _superUserStreamWriter.WriteLine($"/system/bin/screencap {_imgPath}");
            _superUserStreamWriter.Flush();

            // Wait
            // https://stackoverflow.com/a/16160785/5377194
            {
                _superUserStreamWriter.WriteLine("echo -n 0");
                _superUserStreamWriter.Flush();

                _superUser.InputStream.ReadByte();
            }

            using var f = File.OpenRead(_imgPath);
            using var reader = new BinaryReader(f, Encoding.ASCII);
            var w = reader.ReadInt32();
            var h = reader.ReadInt32();
            var format = reader.ReadInt32();

            if (Build.VERSION.SdkInt >= BuildVersionCodes.O)
            {
                reader.ReadInt32();
            }

            if (_buffer == null)
            {
                // If format is not RGBA, notify
                if (format != 1)
                {
                    AutomataApi.Toast($"Unexpected raw image format: {format}");
                }

                _buffer = new byte[w * h * 4];
                _rootLoadMat = new Mat(h, w, CvType.Cv8uc4);
            }

            reader.Read(_buffer, 0, _buffer.Length);

            _rootLoadMat.Put(0, 0, _buffer);

            Imgproc.CvtColor(_rootLoadMat, _rootConvertMat, Imgproc.ColorRgba2gray);

            return new DroidCvPattern(_rootConvertMat, false);
        }

        public void Dispose()
        {
            _rootLoadMat?.Release();
            _rootLoadMat = null;

            _rootConvertMat?.Release();
            _rootConvertMat = null;

            _buffer = null;

            _superUserStreamWriter.WriteLine("exit");
            _superUserStreamWriter.Flush();
            _superUserStreamWriter.Dispose();

            _superUser.WaitFor();
        }
    }
}

File: app/src/main/java/com/mathewsachin/fategrandautomata/root/RootScreenshotService.kt
package com.mathewsachin.fategrandautomata.root

import android.os.Build
import com.mathewsachin.fategrandautomata.imaging.DroidCvPattern
import com.mathewsachin.fategrandautomata.util.readIntLE
import com.mathewsachin.libautomata.ColorManager
import com.mathewsachin.libautomata.Pattern
import com.mathewsachin.libautomata.ScreenshotService
import org.opencv.core.CvType
import org.opencv.core.Mat
import org.opencv.imgproc.Imgproc
import timber.log.Timber
import java.io.DataInputStream

class RootScreenshotService(
    private val superUser: SuperUser,
    private val colorManager: ColorManager
) : ScreenshotService {
    private var reader: DataInputStream = superUser.inStream
    private var buffer: ByteArray? = null

    private var bufferMat: Mat? = null
    private val grayscaleMat = Mat()
    private val grayscalePattern = DroidCvPattern(grayscaleMat, ownsMat = false)
    private val colorMat = Mat()
    private val colorPattern = DroidCvPattern(colorMat, ownsMat = false)

    private fun screenshotIntoBuffer() {
        superUser.writeLine("/system/bin/screencap")

        val w = reader.readIntLE()
        val h = reader.readIntLE()
        val format = reader.readIntLE()

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            reader.readIntLE()
        }

        if (buffer == null) {
            // If format is not RGBA
            if (format != 1) {
                Timber.e("Unexpected raw image format: $format")
            }

            Timber.d("${w}x${h} format=$format")

            buffer = ByteArray(w * h * 4)
            bufferMat = Mat(h, w, CvType.CV_8UC4)
        }

        // "readFully" will wait for the entire data (b.size) to be available in the input stream,
        // however long it takes (in actuality, it just takes a few milliseconds).
        buffer?.let { b -> reader.readFully(b, 0, b.size) }
        bufferMat?.put(0, 0, buffer)
    }

    override fun takeScreenshot(): Pattern {
        screenshotIntoBuffer()

        return if (colorManager.isColor) {
            Imgproc.cvtColor(bufferMat, colorMat, Imgproc.COLOR_RGBA2BGR)

            colorPattern
        } else {
            Imgproc.cvtColor(bufferMat, grayscaleMat, Imgproc.COLOR_RGBA2GRAY)

            grayscalePattern
        }
    }

    override fun close() {
        bufferMat?.release()
        grayscaleMat.release()
        colorMat.release()

        try {
            superUser.close()
        } catch (e: Exception) {
            Timber.e(e, "Error closing super user")
        }

        grayscalePattern.close()
        colorPattern.close()

        buffer = null
    }
}